keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,465 @@
+`--node-ip` to override the auto-detected node IP, but when using
+cloud-controller-manager, this has no effect.
+
+#### `NodeAddressType` semantics
+
+The difference between `InternalIP` and `ExternalIP` node addresses is
+not well-defined except in public clouds, and the two are used
+inconsistently outside of that environment (though for the most part,
+`ExternalIP` is used to mean "involves address rewriting from a public
+IP to a private IP", and `InternalIP` is used for anything else).
+
+### Goals
+
+- Add `Node.Status.Hostname` and `Node.Status.IPs` and deprecate
+  `Node.Status.NodeAddresses`
+
+- Clarify the expectations of cloud providers and
+  cloud-controller-manager with respect to single-stack IPv6 and
+  dual-stack clusters.
+
+  - Do better error handling when, eg, trying to bring up a single-stack
+    IPv6 cluster on a cloud that only supports IPv4.
+
+- Extend the bare-metal node address code to allow for dual-stack
+
+- Clarify the expected behavior of `--node-ip` and extend it to handle
+  dual-stack clusters.
+
+- Clarify the semantics of "internal" vs "external" node IPs and ensure
+  that all cloud providers are using the distinction correctly.
+
+### Non-Goals
+
+- Requiring cloud providers to implement IPv6 / dual-stack.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+The `NodeAddress` of type `NodeHostName` will be moved to a new
+`Node.Status.Hostname` field.
+
+```
+<<[UNRESOLVED hostname ]>>
+Is `Node.Status.Hostname` required or optional? Check existing constraints.
+<<[/UNRESOLVED]>>
+```
+
+The other information in `Node.Status.Addresses` will be moved to a
+new `Node.Status.IPs` array of type `NodeIP`:
```

Comment:
Can we add that the `PodIPs` field from pods with `HostNetwork=true`  will take the value from this field, or is this out of scope?

Reply:
I like that property, but what value do we get from specifying it?

Reply:
> The other information in `Node.Status.Addresses` will be moved to a new `Node.Status.IPs` array of type `NodeIP`

Can you clarify what you mean by "moved", as long as we're still maintaining the status.addresses field in v1?

Reply:
@liggit "moved" as in "moved from the perspective of clients who only use non-deprecated fields"? I can clarify...

Reply:
xref https://github.com/kubernetes/kubernetes/pull/89817